File: android/src/main/kotlin/com/mytiki/sdk/capacitor/Title.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capacitor

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.mytiki.tiki_sdk_android.channel.Channel
import com.mytiki.tiki_sdk_android.trail.Tag
import com.mytiki.tiki_sdk_android.trail.TrailMethod
import com.mytiki.tiki_sdk_android.trail.req.ReqTitle
import com.mytiki.tiki_sdk_android.trail.req.ReqTitleGet
import com.mytiki.tiki_sdk_android.trail.rsp.RspTitle
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.async

class Title(private val channel: Channel) {
    fun create(call: PluginCall) {
        val deferred = MainScope().async {
            val tags: List<Tag> =
                call.getArray("tags").toList<String>().map { tag -> Tag.from(tag) }
            val title: RspTitle = channel.request(
                TrailMethod.TITLE_CREATE,
                ReqTitle(
                    call.getString("ptr")!!,
                    tags,
                    call.getString("description")
                )
            )
            { args ->
                RspTitle.from(args)
            }.await()
            call.resolve(toJS(title))
        }
    }

    fun get(call: PluginCall) {
        val deferred = MainScope().async {
            val title: RspTitle = channel.request(
                TrailMethod.TITLE_GET,
                ReqTitleGet(call.getString("ptr")!!)
            )
            { args ->
                RspTitle.from(args)
            }.await()
            call.resolve(toJS(title))
        }
    }

    companion object {
        fun toJS(title: RspTitle): JSObject {
            val ret = JSObject()
            ret.put("id", title.id)
            ret.put("hashedPtr", title.hashedPtr)
            ret.put("origin", title.origin)
            ret.put("description", title.description)
            val tags: List<String>? = title.tags?.map { tag -> tag.value }
            ret.put("tags", if (tags != null) JSArray(tags) else null)
            ret.put("timestamp", title.timestamp?.time)
            return ret;
        }
    }
}

File: ios/Plugin/Title.swift
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import TikiSdk
import Capacitor

public class Title{
    
    public var tikiSdk = TikiSdk.config()

    @objc func create(_ call: CAPPluginCall) async{
        do{
            let ptr = call.getString("ptr")
            let tags = call.getArray("tags").map{ arr in
                arr.map{ tag in Tag.from(tag: tag as! String) }
            }
            guard ptr != nil, !ptr!.isEmpty, tags != nil, !tags!.isEmpty else{
                call.reject("Please provide PTR and tags. They are required for Title creation.")
                return
            }
            let description = call.getString("description")
            let origin = call.getString("origin")
            let title = try await tikiSdk.trail.title.create(ptr: ptr!, tags: tags!, description: description, origin: origin)
            call.resolve(Title.toJS(title!))
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func get(_ call: CAPPluginCall) async{
        do{
            guard let ptr = call.getString("ptr") else{
                call.reject("Please provide PTR in plugin call.")
                return
            }
            let origin = call.getString("origin")
            let title = try await tikiSdk.trail.title.get(ptr: ptr, origin: origin)
            if(title == nil){
                call.resolve([:])
            }else{
                call.resolve(Title.toJS(title!))
            }
        }catch{
            call.reject(error.localizedDescription)
        }
    }

    @objc func id(_ call: CAPPluginCall) async{
        do{
            guard let id = call.getString("id") else{
                call.reject("Please provide id in plugin call.")
                return
            }
            let title = try await tikiSdk.trail.title.id(id: id)
            if(title == nil){
                call.resolve([:])
            }else{
                call.resolve(Title.toJS(title!))
            }
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    
    public static func toJS(_ title:TitleRecord) -> JSObject{
        var ret = JSObject()
        ret["hashedPtr"] = title.hashedPtr
        ret["description"] = title.description
        ret["id"] = title.id
        ret["tags"] = title.tags.map{tag in tag.toString()}
        ret["origin"] = title.origin
        return ret
    }
}
